proxylib/memcached/text/parser.go

```diff
@@ -0,0 +1,331 @@
+			meta.Keys = [][]byte{}
+			frameLength = linefeed + 2
+			p.watching = true
+		default:
+			log.Error("Could not parse text memcache frame")
+			return proxylib.ERROR, 0
+		}
+		logEntry := &cilium.LogEntry_GenericL7{
+			&cilium.L7LogEntry{
+				Proto: "textmemcached",
+				Fields: map[string]string{
+					"command": meta.Command,
+					"keys":    string(bytes.Join(meta.Keys, []byte(", "))),
+				},
+			},
+		}
+
+		p.requestCount++
+		r := &replyIntent{
+			requestID: p.requestCount,
+			command:   command,
+		}
+
+		matches := p.connection.Matches(meta)
+
+		if matches {
+			r.denied = false
+			p.replyQueue = append(p.replyQueue, r)
+			p.connection.Log(cilium.EntryType_Request, logEntry)
+			return proxylib.PASS, frameLength
+		}
+
+		r.denied = true
+		if !noreply {
+			p.injectDeniedMessage()
+			if p.requestCount == p.replyCount+1 {
+				p.injectDeniedMessage()
+			} else {
+				p.replyQueue = append(p.replyQueue, r)
+			}
+		}
+		p.connection.Log(cilium.EntryType_Denied, logEntry)
+		return proxylib.DROP, frameLength
+	}
+
+	//reply
+	log.Debugf("reply, parsing to figure out if we have it all")
+
+	intent := p.replyQueue[0]
+
+	logEntry := &cilium.LogEntry_GenericL7{
+		&cilium.L7LogEntry{
+			Proto: "textmemcached",
+			Fields: map[string]string{
+				"command": string(intent.command),
+			},
+		},
+	}
+	if p.watching {
+		// in watch mode we pass all replied lines
+		return proxylib.PASS, linefeed + 2
+	}
+
+	switch {
+	case p.isErrorReply(tokens[0]) ||
+		p.isCommandStorage(intent.command) ||
+		p.isCommandDelete(intent.command) ||
+		p.isCommandIncrDecr(intent.command) ||
+		bytes.Equal(intent.command, []byte("touch")) ||
+		bytes.Equal(intent.command, []byte("slabs")) ||
+		bytes.Equal(intent.command, []byte("lru")) ||
+		bytes.Equal(intent.command, []byte("flush_all")) ||
+		bytes.Equal(intent.command, []byte("cache_memlimit")) ||
+		bytes.Equal(intent.command, []byte("version")) ||
+		bytes.Equal(intent.command, []byte("misbehave")):
+
+		// passing one line of reply
+		p.connection.Log(cilium.EntryType_Response, logEntry)
+		return proxylib.PASS, linefeed + 2
+	case p.isCommandRetrieval(intent.command) ||
+		bytes.Equal(intent.command, []byte("stats")):
+		t, nBytes := p.untilEnd(data)
+		if t == proxylib.PASS {
+			p.connection.Log(cilium.EntryType_Response, logEntry)
+			p.replyQueue = p.replyQueue[1:]
+		}
+		return t, nBytes
+	case bytes.Equal(intent.command, []byte("lru_crawler")):
+		// check if it's response line
+		if bytes.Equal(tokens[0], []byte("OK")) ||
+			bytes.Equal(tokens[0], []byte("BUSY")) ||
+			bytes.Equal(tokens[0], []byte("BADCLASS")) {
+			p.connection.Log(cilium.EntryType_Response, logEntry)
+			return proxylib.PASS, linefeed + 2
+		}
+
+		t, nBytes := p.untilEnd(data)
+		if t == proxylib.PASS {
+			p.connection.Log(cilium.EntryType_Response, logEntry)
+			p.replyQueue = p.replyQueue[1:]
+		}
+		return t, nBytes
+	}
+	log.Error("Could not parse text memcache frame")
+	return proxylib.ERROR, 0
```

Comment:
I wonder if it would be better to pass the line in this case?

It seems that some replies are single lines and some require the END. Could we record the fact (e.g., `multiLIine bool` in the reply queue entry and then parse accordingly? That way the reply parser would be simpler and would not depend on the actual command being replied to, and hence there would be no need for the parse error case at all?

Reply:
If the code reaches this line it will probably mean a bug in parser, which would be nice to have it recorded, at least in alpha stage.

`untilEnd` method needs to be optimized for retrieval (`get` command returns a bit more structured reply, but I didn't have time to properly parse this information. There is an issue tracking it already).